pkg/policy/assignment_manager.go

```diff
@@ -0,0 +1,323 @@
+	GetPolicyAssignment(context.Context, *pb.GetPolicyAssignmentRequest) (*pb.PolicyAssignment, error)
+	UpdatePolicyAssignment(context.Context, *pb.PolicyAssignment) (*pb.PolicyAssignment, error)
+	DeletePolicyAssignment(context.Context, *pb.DeletePolicyAssignmentRequest) (*emptypb.Empty, error)
+	ListPolicyAssignments(context.Context, *pb.ListPolicyAssignmentsRequest) (*pb.ListPolicyAssignmentsResponse, error)
+}
+
+type assignmentManager struct {
+	logger       *zap.Logger
+	esClient     esutil.Client
+	esConfig     *config.ElasticsearchConfig
+	indexManager indexmanager.IndexManager
+	filterer     filtering.Filterer
+}
+
+func NewAssignmentManager(
+	logger *zap.Logger,
+	esClient esutil.Client,
+	esConfig *config.ElasticsearchConfig,
+	indexManager indexmanager.IndexManager,
+	filterer filtering.Filterer,
+) AssignmentManager {
+	return &assignmentManager{
+		logger,
+		esClient,
+		esConfig,
+		indexManager,
+		filterer,
+	}
+}
+
+func (m *assignmentManager) CreatePolicyAssignment(ctx context.Context, assignment *pb.PolicyAssignment) (*pb.PolicyAssignment, error) {
+	log := m.logger.Named("CreatePolicyAssignment").
+		With(zap.String("policyVersionId", assignment.PolicyVersionId)).
+		With(zap.String("policyGroup", assignment.PolicyGroup))
+	log.Debug("received request")
+
+	if assignment.PolicyGroup == "" || assignment.PolicyVersionId == "" {
+		return nil, createErrorWithCode(log, "Must specify policy group and a policy version", nil, codes.InvalidArgument)
+	}
+
+	policyId, version, err := parsePolicyVersionId(assignment.PolicyVersionId)
+	if err != nil {
+		return nil, createError(log, "error parsing policy version id", err)
+	}
+
+	if version == 0 {
+		return nil, createErrorWithCode(log, "Assignments must use policy version ids", nil, codes.InvalidArgument)
+	}
+
+	assignmentId := policyAssignmentId(policyId, assignment.PolicyGroup)
+	existingAssignment, err := m.GetPolicyAssignment(ctx, &pb.GetPolicyAssignmentRequest{Id: assignmentId})
+	if existingAssignment != nil {
+		return nil, createErrorWithCode(log, "assignment already exists", nil, codes.AlreadyExists)
+	}
+
+	if err != nil && status.Convert(err).Code() != codes.NotFound {
+		return nil, err
+	}
+
+	response, err := m.esClient.MultiGet(ctx, &esutil.MultiGetRequest{
+		Items: []*esutil.EsMultiGetItem{
+			{
+				Id:      assignment.PolicyVersionId,
+				Index:   m.indexManager.AliasName(constants.PoliciesDocumentKind, ""),
+				Routing: policyId,
+			},
+			{
+				Id:    assignment.PolicyGroup,
+				Index: m.indexManager.AliasName(constants.PolicyGroupsDocumentKind, ""),
+			},
+		},
+	})
+
+	if err != nil {
+		return nil, createError(log, "error retrieving policy version and group", err)
+	}
+
+	for i, resource := range []string{"policy version", "policy group"} {
+		if !response.Docs[i].Found {
+			return nil, createErrorWithCode(log, fmt.Sprintf("%s does not exist", resource), nil, codes.FailedPrecondition)
```

Comment:
might be helpful to include the ID of the resource that wasn't found, like `policy version with ID %s not found`